tools/total-contributions

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+import argparse
+import os
+import pathlib
+import subprocess
+import sys
+from datetime import datetime
+from typing import Dict, List
+
+
+def add_log(committer_dict: Dict[str, int], input: List[str]) -> None:
+    for dataset in input:
+        committer_name = dataset.split("\t")[1]
+        commit_count = int(dataset.split("\t")[0])
+        if committer_name in committer_dict:
+            committer_dict[committer_name] += commit_count
+        else:
+            committer_dict[committer_name] = commit_count
+
+
+def retrieve_log(repo: str, lower_version: str, upper_version: str) -> List[str]:
+    return subprocess.check_output(
+        ["git", "shortlog", "-s", lower_version + ".." + upper_version],
+        cwd=find_path(repo),
+        text=True,
+    ).splitlines()
+
+
+def find_version(time: str, repository: str) -> str:
+    """Find the latest release version for the target repository as of the
+    specified time.
+    """
+    versions_list = subprocess.check_output(
+        ["git", "tag", "-l"], cwd=find_path(repository), text=True
+    ).splitlines()
+    for version in versions_list:
+        version_time = subprocess.check_output(
+            ["git", "log", "-1", "--format=%ai", version],
+            cwd=find_path(repository),
+            text=True,
+        ).split()[0]
+        if datetime.strptime(version_time, "%Y-%m-%d") >= datetime.strptime(time, "%Y-%m-%d"):
+            return version
+    return version
```

Comment:
It looks like this doesn't get the tags sorted by date.

Here's a command that gets them sorted by date descending, and also uses one command instead of N:
```
$ git tag --format '%(creatordate:iso-local) %(refname:lstrip=2)' --sort=-creatordate
```
Or for a more machine-readable version:
```
$ git tag --format '%(creatordate:unix) %(refname:lstrip=2)' --sort=-creatordate
```

Reply:
(And then it should take the first entry it finds that's *before* the given time, rather than after)

Reply:
Applied the `--sort`; is the `--format` here required?

Reply:
The point of the `--format` is that it prints the dates as well as the names, so that the job requires only 1 command instead of N + 1. For example in zulip-mobile:
```
$ git tag --format '%(creatordate:iso-local) %(refname:lstrip=2)' --sort=-creatordate | head -3
2022-03-16 13:34:05 -0700 v27.182
2022-02-11 15:30:47 -0800 v27.181
2022-01-25 16:59:10 -0800 v27.180
```

There's a couple more refinements reflected there. One is that the `%ci` in the existing version (`%ai` above) will print times in whatever time zone the committer was using. That can throw the date off by one. Saying `creatordate:iso` would reproduce that bug, but `creatordate:iso-local` fixes it, by using a consistent timezone (namely your own as you're running the command.)

And then saying `creatordate:unix` goes further by just making it a numeric timestamp, so there's no messy having to parse a date, or having to choose between parsing times and rounding to a day.